### ISSUE TITLE
bleadperl / hex ignores variable length and/or tr doesn't null    terminate

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3806]
Migrated from [rt.perl.org#6728](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=6728) (status was 'resolved')

Searchable as RT6728$

**Comment by p5pRT** [comment]
### From dcd@tc.fluke.com




&nbsp;  \(I've included a patch below that should fix both hex\(\) and oct\(\) \)

The following one liner prints 257 when I 
would expect it to print 1\.  Notice that the argument
to hex prints as "0001" in all cases\.

perl \-wle '$\_="R0\_001";tr/R\_//d;print;print hex';
0001
257


perl \-wle '$\_="0001";tr/R\_//d;print;print hex';
0001
1

perl \-wle '$\_="R0\_001";s/\[R\_\]//g;print ;print hex';
0001
1


dd&#8203;:dcd$ perl \-MDevel&#8203;::Peek \-wle '$\_="R0\_001";tr/R\_//d;Dump $\_;print hex';
SV = PV\(0x81207e4\) at 0x81204b8
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x8132558 "0001"
&nbsp;  CUR = 4
&nbsp;  LEN = 7
257
dd&#8203;:dcd$ perl \-MDevel&#8203;::Peek \-wle '$\_="0001";tr/R\_//d;Dump $\_;print hex';
SV = PV\(0x81207e4\) at 0x81204b8
&nbsp;  REFCNT = 1
&nbsp;  FLAGS = \(POK\,pPOK\)
&nbsp;  PV = 0x8132558 "0001"\\0
&nbsp;  CUR = 4
&nbsp;  LEN = 5
1

Using gdb with a breakpoint in Perl\_scan\_hex
one can see that the hex value being processed 
is "000101"

Perl\_scan\_hex doesn't know the length of the
string it is scanning\, as Perl\_pp\_hex \(\) at pp\.c&#8203;:2674
passes in the constant 99

One may wonder if the tr command should have null terminated
the string\, but Devel&#8203;::Peek Dump prints "0001" as expected
If I use s/\[R\_\]//g instead of tr/R\_//d the value return
from hex is printed correctly\.

So either tr should null terminate\, or pp\_hex should pay
attention to the length\, or perhaps both

perlguts states&#8203;:
&nbsp;   "All SVs that contain strings should be terminated with a NUL character\."
which leads me to think that tr should be fixed
but the "typical" meantioned just a few sentances later\,
&nbsp;   "Perl's own functions typically add a trailing NUL for this reason\.
&nbsp;   Nevertheless\, you should be very careful when you pass a string stored
&nbsp;   in an SV to a C function or system call\."
leads me to believe that this may be optional\, so hex should do
a better job of checking\.

There may be a similar problem in oct\(\)

\./miniperl \-wle '$\_="0xR0\_001";tr/R\_//d;print;print oct';
0x0001
257

\./miniperl \-wle '$\_="0R0\_001";tr/R\_//d;print;print oct';
00001
65

\./miniperl \-wle '$\_="0bR0\_001";tr/R\_//d;print;print oct';
0b0001
5

I didn't add any new test cases\, as I wouldn't want to
force tr///d to continue to not terminate strings\,
but here is a patch that fixes the above cases\,
and still passes make test\.

<details><summary>Inline Patch</summary>

```diff
--- pp.c.save	Wed Mar 28 07:14:43 2001
+++ pp.c	Wed Apr  4 12:40:14 2001
@@ -2667,11 +2667,11 @@
     dSP; dTARGET;
     char *tmps;
     STRLEN argtype;
-    STRLEN n_a;
+    STRLEN len;
 
-    tmps = POPpx;
+    tmps = (SvPVx(POPs, len));
     argtype = 1;		/* allow underscores */
-    XPUSHn(scan_hex(tmps, 99, &argtype));
+    XPUSHn(scan_hex(tmps, len, &argtype));
     RETURN;
 }
 
@@ -2681,20 +2681,20 @@
     NV value;
     STRLEN argtype;
     char *tmps;
-    STRLEN n_a;
+    STRLEN len;
 
-    tmps = POPpx;
-    while (*tmps && isSPACE(*tmps))
-	tmps++;
+    tmps = (SvPVx(POPs, len));
+    while (*tmps && len && isSPACE(*tmps))
+	tmps++, len--;
     if (*tmps == '0')
-	tmps++;
+	tmps++, len--;
     argtype = 1;		/* allow underscores */
     if (*tmps == 'x')
-	value = scan_hex(++tmps, 99, &argtype);
+	value = scan_hex(++tmps, --len, &argtype);
     else if (*tmps == 'b')
-	value = scan_bin(++tmps, 99, &argtype);
+	value = scan_bin(++tmps, --len, &argtype);
     else
-	value = scan_oct(tmps, 99, &argtype);
+	value = scan_oct(tmps, len, &argtype);
     XPUSHn(value);
     RETURN;
 }
```
</details>


<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=high

This perlbug was built using Perl v5.7.0 - Fri Jan 12 16:03:35 PST 2001
It is being executed now by  Perl v5.7.0 - Wed Apr  4 10:54:02 PDT 2001.

Site configuration information for perl v5.7.0:

Configured by dcd at Wed Apr  4 10:54:02 PDT 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.3-ac2, archname=i686-linux
    uname='linux dd 2.4.3-ac2 #2 tue apr 3 23:44:34 pdt 2001 i686 '
    config_args='-Dinstallusrbinperl -Dusedevel -Doptimize=-O3 -g -de -Dcf_email=dcd@tc.fluke.com'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3 -g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='egcs-2.91.66.1 19990314/Linux (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lgdbm -ldbm -ldb -ldl -lm -lc
    perllibs=-ldl -lm -lc
    libc=/lib/libc.so.5.4.44, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL8403


@INC for perl v5.7.0:
    /usr/local/lib/perl5/5.7.0/i686-linux
    /usr/local/lib/perl5/5.7.0
    /usr/local/lib/perl5/site_perl/5.7.0/i686-linux
    /usr/local/lib/perl5/site_perl/5.7.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.0:
    HOME=/home/dcd
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/dcd/bin:/sbin:/usr/local/bin:/bin:/usr/bin:/usr/X11/bin:/usr/games:/usr/local/samba:/home/hobbes/tools/scripts:/home/hobbes/tools/linux:/usr0/hobbes/tools/scripts:/usr0/dcd/bin:/apps/general/bin:/usr/public
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

and my fix was to correct pp\_hex and pp\_oct\, and Gisle added
> tr is obviously buggy\.  I don't think pp\_hex should need to change\,
> but since you already done it is probably a good thing\.

The code in the hex and oct implementations in doop\.c stands
out in it arbitrary limit of 99\.

This lead me to the see that something still should be done
to fix hex and oct since even back in perl 5\.6\.0 the limit
of 99 can break some inputs to oct and hex

$ perl5\.6\.0 \-le 'print oct "0b10"'
2
$ perl5\.6\.0 \-le 'print oct "0b"\.\("0"x10\)\."10"'
2
$ perl5\.6\.0 \-le 'print oct "0b"\.\("0"x100\)\."10"'
0
&nbsp;   \#\# the above should have printed 2

$ perl5\.6\.0 \-le 'print hex "10"'
16
$ perl5\.6\.0 \-le 'print hex \+\("0"x10\)\."10"'
16
$ perl5\.6\.0 \-le 'print hex \+\("0"x100\)\."10"'
0
&nbsp;   \#\# the above should have printed 16


The following patch to t/op/oct\.t will show how perls without
the patch that I supplied to fix 20010404\.009 will fail\.

<details><summary>Inline Patch</summary>

```diff
--- t/op/oct.t.save	Thu Oct 12 18:30:35 2000
+++ t/op/oct.t	Thu Apr  5 09:08:09 2001
@@ -1,6 +1,6 @@
 #!./perl

-print "1..44\n";
+print "1..50\n";

 print +(oct('0b1_0101') ==        0b101_01) ? "ok" : "not ok", " 1\n";
 print +(oct('0b10_101') ==           0_2_5) ? "ok" : "not ok", " 2\n";
@@ -77,3 +77,12 @@
 else {
     print "\x2F_" eq "/_"                  ? "ok" : "not ok", " 44\n";
 }
+
+print +(oct('0b'.(  '0'x10).'1_0101') ==  0b101_01) ? "ok" : "not ok", " 45\n";
+print +(oct('0b'.( '0'x100).'1_0101') ==  0b101_01) ? "ok" : "not ok", " 46\n";
+print +(oct('0b'.('0'x1000).'1_0101') ==  0b101_01) ? "ok" : "not ok", " 47\n";
+
+print +(hex((  '0'x10).'01234') ==  0x1234) ? "ok" : "not ok", " 48\n";
+print +(hex(( '0'x100).'01234') ==  0x1234) ? "ok" : "not ok", " 49\n";
+print +(hex(('0'x1000).'01234') ==  0x1234) ? "ok" : "not ok", " 50\n";
+
```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

David Dyck \<dcd@&#8203;tc\.fluke\.com> writes&#8203;:

>  dd&#8203;:dcd$ perl \-MDevel&#8203;::Peek \-wle '$\_="R0\_001";tr/R\_//d;Dump $\_;print hex';
>  SV = PV\(0x81207e4\) at 0x81204b8
>    REFCNT = 1
>    FLAGS = \(POK\,pPOK\)
>    PV = 0x8132558 "0001"
>    CUR = 4
>    LEN = 7
>  257

\[\.\.\.\]

>  So either tr should null terminate\, or pp\_hex should pay
>  attention to the length\, or perhaps both

tr is obviously buggy\.  I don't think pp\_hex should need to change\,
but since you already done it is probably a good thing\.

The following patch fixes tr&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- doop.c.dist	Wed Apr  4 14:42:01 2001
+++ doop.c	Wed Apr  4 14:42:28 2001
@@ -184,6 +184,7 @@
 		s++;
 	    }
 	}
+	*d = '\0';
 	SvCUR_set(sv, d - dstart);
     }
     else { /* isutf8 */

Regards,
Gisle
```
</details>
